token-erc-20/chaincode-java/src/main/java/org/hyperledger/fabric/samples/erc20/ERC20TokenContract.java

```diff
@@ -0,0 +1,444 @@
+    String balanceBytes = stub.getStringState(balanceKey.toString());
+    if (stringIsNullOrEmpty(balanceBytes)) {
+      String errorMessage = String.format("The account  %s does not exist", clientAccountID);
+      throw new ChaincodeException(errorMessage, NOT_FOUND.toString());
+    }
+    return Long.parseLong(balanceBytes);
+  }
+
+  /**
+   * ClientAccountID returns the id of the requesting client's account. In this implementation, the
+   * client account ID is the clientId itself. Users can use this function to get their own account
+   * id, which they can then give to others as the payment address.
+   *
+   * @return client account id .
+   */
+  @Transaction(intent = Transaction.TYPE.EVALUATE)
+  public String ClientAccountID(final Context ctx) {
+    // Get ID of submitting client identity
+    return ctx.getClientIdentity().getId();
+  }
+
+  /**
+   * Return the total token supply.
+   *
+   * @param ctx the transaction context
+   * @return the total token supply
+   */
+  @Transaction(intent = Transaction.TYPE.EVALUATE)
+  public long TotalSupply(final Context ctx) {
+    String totalSupply = ctx.getStub().getStringState(TOTAL_SUPPLY_KEY.getValue());
+    if (stringIsNullOrEmpty(totalSupply)) {
+      throw new ChaincodeException("Total Supply  not found", NOT_FOUND.toString());
+    }
+    return Long.parseLong(totalSupply);
+  }
+
+  /**
+   * Allows `spender` to spend `value` amount of tokens from the owner.
+   *
+   * @param ctx the transaction context
+   * @param spender The spender
+   * @param value The amount of tokens to be approved for transfer
+   */
+  @Transaction(intent = Transaction.TYPE.SUBMIT)
+  public void Approve(final Context ctx, final String spender, final long value) {
+    ChaincodeStub stub = ctx.getStub();
+    String owner = ctx.getClientIdentity().getId();
+    CompositeKey allowanceKey =
+        stub.createCompositeKey(ALLOWANCE_PREFIX.getValue(), owner, spender);
+    stub.putStringState(allowanceKey.toString(), String.valueOf(value));
+    Approval approval = new Approval(owner, spender, value);
+    stub.setEvent(APPROVAL.getValue(), approval.toJSONString().getBytes(UTF_8));
+  }
+
+  /**
+   * Returns the amount of tokens which `spender` is allowed to withdraw from `owner`.
+   *
+   * @param ctx the transaction context
+   * @param owner The owner of tokens
+   * @param spender The spender who are able to transfer the tokens
+   * @return the amount of remaining tokens allowed to spent
+   */
+  @Transaction(intent = Transaction.TYPE.SUBMIT)
+  public long Allowance(final Context ctx, final String owner, final String spender) {
+    ChaincodeStub stub = ctx.getStub();
+    CompositeKey allowanceKey =
+        stub.createCompositeKey(ALLOWANCE_PREFIX.getValue(), owner, spender);
+    String allowanceBytes = stub.getStringState(allowanceKey.toString());
+    if (stringIsNullOrEmpty(allowanceBytes)) {

```

Comment:
I think this should not be an exception, but should return a zero-valued Allowance.   The golang erc20 does: 
```
	// If no current allowance, set allowance to 0
	if allowanceBytes == nil {
		allowance = 0
	} else {
		allowance, err = strconv.Atoi(string(allowanceBytes)) // Error handling not needed since Itoa() was used when setting the totalSupply, guaranteeing it was an integer.
	}
```

Reply:
Fixed the code to return zero .